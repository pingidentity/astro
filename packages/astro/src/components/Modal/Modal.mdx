<Meta title="Components/Modal/Modal" />

# Modal

Modals (also known as modal windows, overlays, and dialogs) are large UI elements that sit on top of an application's main window.
To return to the application's main interface, users must interact with the modal.

This component must be wrapped in an OverlayProvider and the first child component should be a trigger, such as a button.

If the modal is used to display an error message, it displays after users perform an action and require a response. 
These messages should be short, meaningful, free of technical jargon, and follow status color guidelines. 
They can also be used to educate users, but don’t try to explain complicated troubleshooting processes in error messages.

Error messages should clearly explain: 
- What happened
- Why it happened, if that’s relevant
- The steps users should take to resolve the issue

See our [Error message writing guidelines](https://pingidentity.atlassian.net/wiki/spaces/CCC/pages/21137273/Error+messages) for additional information. 

Modals should:
- Be used for important warnings and information requests as a way to prevent or correct critical errors.
- Be used to break down complex workflow to not overwhelm the user.
- Have all the critical information inside it to help users make decisions.

Modals should not:
- Be nested inside or replaced by another modal.
- Be overused as it disrupts the user’s workflow and requires a response.
- Have content that requires scrolling. If the content length is exceedingly large or requires complex forms, use the [OverlayPanel](./?path=/docs/components-overlaypanel--default) component.
- Be used for brief warnings and alerts, use the [Messages](./?path=/docs/components-messages--default) component instead.

### Required Components

This component requires the OverlayProvider and [Button](./?path=/docs/components-button--default) components. 

### Accessibility

This component should adhere to the [WAI-ARIA Modal](https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/examples/dialog/) accessibility guidelines.

#### Keyboard Navigation

These keys provide additional functionality to the component. 

| Key | Functions |
| ---- | ---- |
| Tab | The trigger component and the content in the modal are focusable using the Tab key.   |
|  | When a modal opens, focus moves to an element contained in the modal, which is usually the first focusable element. If the first focusable element does not make sense, tabindex=”-1” should be added to a static element at the top of the modal, which could be the panel title. |
| Space or Enter | Opens the modal when the trigger is focused. Once opened, the focus is locked inside the modal. |
| Shift + Tab | Moves focus to the previous focusable component. |
| Esc | Pressing the escape key closes the modal and adds focus to the previously focused component. |

#### Screen Readers

This component uses the following attributes to assist screen readers:
- The button uses the **`aria-label`** attribute to provide an accessible name.
- The **`aria-labelledby`** attribute is supplied with label ID.
- The **`aria-modal`** attribute allows the assistive technologies to inform users that the ability to interact with or access other content on the page requires the modal dialog to be closed.