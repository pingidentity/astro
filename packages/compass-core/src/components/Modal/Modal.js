import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import { Portal } from 'react-portal';
import { noop } from 'underscore';
import FocusTrap from 'focus-trap-react';
import { missingDependencyMessage } from '../../utils/DependencyUtils';
import { useClickOutsideHandler } from '../../utils/MouseUtils';
import { useEscapeKeyHandler } from '../../utils/KeyboardUtils';

missingDependencyMessage(Portal, 'react-portal', 'React Portal');

const Modal = ({
    onOutsideClick,
    onEscapeKey,
    ...props
}) => {
    const modalRef = useRef(null);
    useClickOutsideHandler([modalRef], onOutsideClick);
    useEscapeKeyHandler(onEscapeKey);

    return (
        <Portal>
            <FocusTrap
                focusTrapOptions={{
                    escapeDeactivates: false,
                    fallbackFocus: () => modalRef.current,
                }}
            >
                <div><div tabIndex="-1" {...props} ref={modalRef} /></div>
            </FocusTrap>
        </Portal>
    );
    // there are two divs because FocusTrap and useClickOutsideHandler don't play together otherwise
};

Modal.propTypes = {
    /** Event that is fired when a click is detected outside of the modal */
    onOutsideClick: PropTypes.func,
    /** Event that is fired when the escape key is pressed */
    onEscapeKey: PropTypes.func,
};

Modal.defaulProps = {
    onOutsideClick: noop,
};

export default Modal;
