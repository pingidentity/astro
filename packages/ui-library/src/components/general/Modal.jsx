"use strict";

var React = require("re-react"),
    Utils = require("../../util/Utils"),
    EventUtils = require("../../util/EventUtils.js"),
    If = require("./If.jsx"),
    classnames = require("classnames");

/**
 * @enum {string}
 * @alias Modal.Type
 **/
var Type = {
    BASIC: "basic",
    DIALOG: "dialog",
    ALERT: "alert"
};

/**
* @callback Modal~onClose
*/

/**
 * @class Modal
 * @desc A generic modal component. Child components of the modal are rendered as the modal content.
 *
 * @param {string} [data-id="modal-button"]
 *     To define the base "data-id" value for top-level HTML container.
 * @param {string} [id]
 *     DEPRECATED. Use "data-id" instead. To define the base "data-id" value for top-level HTML container.
 * @param {string} [className]
 *     CSS classes to set on the top-level HTML container.
 * @param {boolean} [expanded=false]
 *     Whether the modal is expanded or not.
 * @param {string} [modalTitle]
 *     Title of the modal.
 * @param {boolean} [showHeader=true]
 *     Controls modal header rendering; if set to false,
 *     making modal effectivly a 'light box' for previews.
 * @param {Modal~onClose} [onClose]
 *     Callback to be triggered when the modal is closed by clicking the close modal link.
 *     If this function returns false then closing will be prevented.
 *     This function is set on the context as "close", for modal children (eg. the "close" button
 *     on the modal) to be able to close the modal by calling the context function.
 * @param {boolean} [maximize=false]
 *     When true, modal content will always occupy the maximum modal dimensions.
 * @param {Modal.Type} [type=Modal.Type.BASIC]
 *     Basic modal when not specified; a DIALOG modal has the "dialog" CSS class on it,
 *     same goes for the ALERT dialog.
 *
 * @example
 *      <a onClick={this._toggleModal}>Open the Modal</a>
 *      <Modal modalTitle="My wonderful modal" onClose={this._toggleModal} expanded={this.state.expanded}>
 *          <p>Thank you for opening this modal.</p>
 *      </Modal>
 */
var Modal = React.createClass({

    propTypes: {
        "data-id": React.PropTypes.string.affectsRendering,
        id: React.PropTypes.string.affectsRendering,
        className: React.PropTypes.string.affectsRendering,
        expanded: React.PropTypes.bool.affectsRendering,
        modalTitle: React.PropTypes.string.affectsRendering,
        showHeader: React.PropTypes.bool.affectsRendering,
        onClose: React.PropTypes.func,
        maximize: React.PropTypes.bool.affectsRendering,
        type: React.PropTypes.oneOf([
            Type.BASIC,
            Type.DIALOG,
            Type.ALERT
        ]).affectsRendering,
        children: React.PropTypes.oneOfType([
            React.PropTypes.object,
            React.PropTypes.arrayOf(React.PropTypes.object)
        ]).affectsRendering
    },

    childContextTypes: {
        close: React.PropTypes.func
    },

    getDefaultProps: function () {
        return {
            "data-id": "modal-button",
            expanded: false,
            showHeader: true,
            maximize: false,
            type: Type.BASIC
        };
    },

    /*
     * Set close method into context to allow children
     * to easily close modal.
     */
    getChildContext: function () {
        return {
            close: this.props.onClose
        };
    },

    _handleKeyDown: function (e) {
        if (!this.props.expanded) {
            return;
        }

        EventUtils.callIfOutsideOfContainer(this.refs.container, function () {
            e.stopPropagation();
            e.preventDefault();
        }, e);
    },

    _getCloseButton: function () {
        return (
            <a data-id="close-button" className="close-modal" onClick={this.props.onClose}>Ã—</a>
        );
    },

    componentWillMount: function () {
        if (this.props.id) {
            console.warn(Utils.deprecateMessage("id", "data-id"));
        }
    },

    componentDidMount: function () {
        window.addEventListener("keydown", this._handleKeyDown);
    },

    componentWillUnmount: function () {
        window.removeEventListener("keydown", this._handleKeyDown);
    },

    render: function () {
        if (!this.props.expanded) {
            return null;
        }

        var dataId = this.props.id || this.props["data-id"];
        var modalClasses = classnames(
            "modal",
            this.props.className,
            {
                maximize: this.props.maximize,
                dialog: this.props.type === Type.DIALOG,
                alert: this.props.type === Type.ALERT,
                show: this.props.expanded
            }
        );

        return (
            <div data-id={dataId} ref="container" key="modal" className={modalClasses}>
                <div className="modal-bg">{/* applies overlay to content area */}</div>
                <div className="modal-content">
                    <If test={this.props.showHeader}>
                        <div className="modal-header">
                            {this.props.modalTitle}
                            {this._getCloseButton()}
                        </div>
                    </If>
                    <div className="modal-body">
                        <If test={!this.props.showHeader}>
                            {this._getCloseButton()}
                        </If>
                        {this.props.children}
                    </div>
                </div>
            </div>
        );
    }
});

Modal.Type = Type;

module.exports = Modal;
