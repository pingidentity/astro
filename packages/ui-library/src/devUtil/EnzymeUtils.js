import { mount } from "enzyme";


export const snapshotDataIds = node => {
    const elements = node.find("[data-id]").hostNodes(); // Find only actual HTML elements

    const ids = elements.map(element => element.prop("data-id")).filter(item => item !== "");
    expect(ids).toMatchSnapshot();
};

export const mountSnapshotDataIds = node => snapshotDataIds(mount(node));
