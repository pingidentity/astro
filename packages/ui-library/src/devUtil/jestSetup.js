
// This is needed for testing the modal and DetailsTooltip component.
global.getSelection = jest.fn();
global.getSelection.mockReturnValue({
    toString: () => "",
});
global.document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
        nodeName: "BODY",
        ownerDocument: document,
    },
});
