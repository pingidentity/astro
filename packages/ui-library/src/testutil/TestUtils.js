/*eslint-disable valid-jsdoc*/

"use strict";

/**
 * @module util/TestUtils
 * @desc A collection of ReactJS test utils, to complement the ones provided by the
 * framework. This is a copy of the TestUtils in the library, just because this class
 * is used in the tests and should work with the same version of React used in the WP
 * and not in the library.
 */

var ReactDOM = require("react-dom");
var React = require("react");
var ReactTestUtils = require("react-addons-test-utils");
var _ = require("underscore");
var assign = require("object-assign");

var TestUtils = {

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithDataId
     * @desc Return all DOM nodes children of the given tree with the specified dataId.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} dataId - the value of the data-id attribute on the target elements
     * @return {array} - of DOM node objects which match the given criteria
     */
    scryRenderedDOMNodesWithDataId: function (parent, dataId) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.getAttribute("data-id") === dataId) {
            nodes = [parentNode];
        }
        else {
            var selectorValue = dataId
                .replace(/\[/g, "\\[")
                .replace(/\]/g, "\\]")
                .replace(/'/g, "\\'");
            nodes = parentNode.querySelectorAll("[data-id=" + selectorValue + "]");
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithDataId
     * @desc Return the single DOM node child of the given tree with the specified dataId.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} dataId - the value of the data-id attribute on the target element
     * @return {object} - the DOM node object which matches the given criteria
     */
    findRenderedDOMNodeWithDataId: function (parent, dataId) {
        var nodes = this.scryRenderedDOMNodesWithDataId(parent, dataId);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with dataId: " +
                dataId + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithTag
     * @desc Return all DOM nodes children of the given tree with the specified tag name.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} tagName - the tag name on the target elements
     * @return {array} - array of DOM nodes object which match the given criteria
     */
    scryRenderedDOMNodesWithTag: function (parent, tagName) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.tagName.toLowerCase() === tagName.toLowerCase()) {
            nodes = [parentNode];
        }
        else {
            nodes = parentNode.querySelectorAll(tagName);
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithTag
     * @desc Return the single DOM node child of the given tree with the specified tag name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} tagName - the tag name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithTag: function (parent, tagName) {
        var nodes = this.scryRenderedDOMNodesWithTag(parent, tagName);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with tag name: " +
                tagName + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithName
     * @desc Return all DOM nodes children of the given tree with the specified name.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} name - the name on the target elements
     * @return {DOMNode[]} - all DOM nodes which match the given criteria
     */
    scryRenderedDOMNodesWithName: function (parent, name) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.name === name) {
            nodes = [parentNode];
        }
        else {
            nodes = parentNode.querySelectorAll("[name='" + name.replace("'", "\\'") + "']");
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithName
     * @desc Return the single DOM node child of the given tree with the specified name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} name - the name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithName: function (parent, name) {
        var nodes = this.scryRenderedDOMNodesWithName(parent, name);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with name: " +
                name + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithClass
     * @desc Return all DOM nodes children of the given tree with the specified CSS class name.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} className - the CSS class name on the target elements (or a concatenation of classes, separated by space)
     * @return {DOMNode[]} - all DOM nodes which match the given criteria
     */
    scryRenderedDOMNodesWithClass: function (parent, className) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode = ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
            ? ReactDOM.findDOMNode(parent)
            : parent;

        var nodes;

        // first check if the root element has the provided classes;
        // if not, build a selector and look for any children

        var parentMatches = _.every(className.split(" "), function (name) {
            return (" " + parentNode.className + " ").indexOf(" " + name + " ") > -1;
        });
        if (parentMatches) {
            nodes = [parentNode];
        }
        else {
            var selector = _.reduce(className.split(" "), function (memo, name) {
                return memo + "." + name;
            }, "");
            nodes = parentNode.querySelectorAll(selector);
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithClass
     * @desc Return the single DOM node child of the given tree with the specified CSS class name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} className - the CSS class name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithClass: function (parent, className) {
        var nodes = this.scryRenderedDOMNodesWithClass(parent, className);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with class name: " +
                className + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedComponentsWithType
     * @desc Finds all instances of components with type equal to `componentType`.
     * @param {object} root - a composite (ie. React) component
     * @param {object} componentType - the component type
     * @return {array} an array of all the matches.
     */
    scryRenderedComponentsWithType: function (root, componentType) {
        return ReactTestUtils.scryRenderedComponentsWithType(root, componentType);
    },

    /**
     * @alias module:util/TestUtils.findRenderedComponentWithType
     * @desc Same as `scryRenderedComponentsWithType` but expects there to be one result
     * and returns that one result, or throws exception if there is any other
     * number of matches besides one.
     * @param {object} root - a composite (ie. React) component
     * @param {object} componentType - the component type
     * @return {!ReactComponent} The one match.
     */
    findRenderedComponentWithType: function (root, componentType) {
        return ReactTestUtils.findRenderedComponentWithType(root, componentType);
    },

    /**
     * @alias module:util/TestUtils.captureGlobalListener
     * @desc Attaches proxy function for `window.addEvenListener` which will save reference to actual provided callback when
     * event type match filter. Useful for testing ReactJS component global (window) callbacks behaviour.
     *
     * @param {string} eventType event to save callback for. case insensitive.
     * @param {object} [obj=window] optional object to capture listener from. Like `document`. If not provided - window object
     * @returns {function} - when executed will trigger actual provided callback to `window.addEventListener`
     *
     * @example
     * //capture registered callback from reactjs control
     * var onKeyDown = TestUtils.captureGlobalListener("keyDown");
     *
     * onKeyDown({keyCode:13});  //later in test trigger global event
     *
     * @example
     * //capture registered callback from reactjs control on `document` object
     * var onKeyDown = TestUtils.captureGlobalListener("keyDown", document);
     *
     * onKeyDown({keyCode:13});  //later in test trigger global event
     */
    captureGlobalListener: function (eventType, obj) {

        var captured;

        var target = obj || window;

        target.addEventListener = function (event, listener) {
            if (event.toLowerCase() === eventType.toLowerCase()) {
                captured = listener;
            }
        };

        return function () {
            return captured.apply(this, arguments);
        };
    }

    /**
     * @desc Return all components in the supplied tree with the specified dataId.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} dataId - the value of the data-id attribute on the target elements
     * @return {ReactComponent[]} - all components which match the given criteria
     */
    /*
     scryRenderedDOMComponentsWithDataId: function (tree, dataId) {
     var components = ReactTestUtils.findAllInRenderedTree(tree, function (inst) {
     var instId = ReactDOM.findDOMNode(inst).getAttribute("data-id");
     return ReactTestUtils.isDOMComponent(inst) && (instId === dataId);
     });

     return components;
     },
     */

    /**
     * @desc Return the single component in the supplied tree with the specified dataId.
     * If no component in the tree matches the criteria, null is returned.
     * If there are more than one component, an error is thrown.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} dataId - the value of the data-id attribute on the target element
     * @return {ReactComponent} - the component which match the given criteria
     * @throws Will throw an Error if more than one element match the criteria
     */
    /*
     findRenderedDOMComponentWithDataId: function (tree, dataId) {
     var components = TestUtils.scryRenderedDOMNodesWithDataId(tree, dataId);

     if (components.length > 1) {
     throw new Error("Found more than one component with dataId: " +
     dataId + " in the given tree");
     }

     return components.length === 0 ? null : components[0];
     },
     */

    /**
     * @desc Return the single component in the supplied tree with the specified name.
     * If no component in the tree matches the criteria, null is returned.
     * If there are more than one component, an error is thrown.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} name - the value of the name attribute on the target element
     * @return {ReactComponent} - the component which match the given criteria
     * @throws Will throw an Error if more than one element match the criteria
     */
    /*
     findRenderedDOMComponentWithName: function (tree, name) {
     var components = ReactTestUtils.findAllInRenderedTree(tree, function (inst) {
     var instName = ReactDOM.findDOMNode(inst).name;
     return ReactTestUtils.isDOMComponent(inst) && (name === instName);
     });

     if (components.length > 1) {
     throw new Error("Found more than one component with name: " +
     name + " in the given tree");
     }

     return components.length === 0 ? null : components[0];
     }
     */
};

/**
 * @class UpdatePropsWrapper
 *
 * @desc
 * A simple wrapper to be used for unit testing, when you need to update
 * a property on a rendered component. Starting with React 0.14.x,
 * directly setting or replacing a prop on a rendered component is deprecated.
 * Instead, use this wrapper and call _setProps(newProps) on it
 * to have the wrapped component re-render with the new property(ies).
 * The ref attribute on the wrapper is set to "wrapper".
 *
 * @Example
 *  // initial render
 *  var component = ReactTestUtils.renderIntoDocument(
 *      <UpdatePropsWrapper type={If} test={true} />
 *  );
 *  // update the property through re-render
 *  component._setProps({test: false});
 *
 * @Example
 *  var component = ReactTestUtils.renderIntoDocument(
 *      <UpdatePropsWrapper type={TabbedSections}
 *              onSectionChange={jest.genMockFunction()}
 *              selectedIndex={-1}>
 *          <div data-id="section1" title="section 1">section 1</div>
 *          <div data-id="section2" title="section 2">section 2</div>
 *      </UpdatePropsWrapper>
 *  );
 *  component._setProps({ selectedIndex: 0 });
 *
 */
TestUtils.UpdatePropsWrapper = React.createClass({
    /**
     * Re-render the wrapped component using the new property.
     * Use this function to avoid setting the props directly on the component (anti pattern).
     * @param {object} props - new property(ies) to set
     */
    _setProps: function (props) {
        this.setState(props);
    },
    getInitialState: function () {
        return assign({ ref: "wrapper" }, this.props);
    },
    render: function () {
        return React.createElement(
            this.props.type, // this is the type of element to create, provided as prop
            this.state // this is the rest of props to set on the element
        );
    }
});

module.exports = TestUtils;
