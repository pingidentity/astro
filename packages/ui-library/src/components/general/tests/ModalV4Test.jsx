jest.mock("popper.js");
jest.mock("react-portal");

jest.dontMock("../Modal");
jest.dontMock("../If");
jest.dontMock("../../../util/EventUtils.js");
jest.dontMock("../../tooltips/CancelTooltip");
jest.dontMock("../../tooltips/DetailsTooltip");
jest.dontMock("../../../util/Utils");

import React from "react";
import ReactDOM from "react-dom";
import { mountSnapshotDataIds } from "../../../devUtil/EnzymeUtils";
import ReactTestUtils from "react-dom/test-utils";
import TestUtils from "../../../testutil/TestUtils";
import Modal, { BodyTitle } from "../Modal";
import { shallow } from "enzyme";
import _ from "underscore";
const Wrapper = TestUtils.UpdatePropsWrapper;

describe("ModalTest v4", function () {

    window.addEventListener = jest.fn();
    window.removeEventListener = jest.fn();
    window.setTimeout = jest.fn();
    beforeEach(function () {
        window.addEventListener.mockClear();
        window.removeEventListener.mockClear();
        window.setTimeout.mockClear();
    });

    function getComponent (opts) {
        const modalDefaults = _.defaults(opts || {}, {
            onOpen: jest.fn(),
            onClose: jest.fn().mockReturnValue(true),
        });

        return ReactTestUtils.renderIntoDocument(<Modal {...modalDefaults} />);
    }

    it("data-id's don't change", () => {
        mountSnapshotDataIds(
            <Modal
                expanded
            />
        );
    });

    it("detaches event listeners and calls close event on unmount", function () {
        const component = getComponent({ expanded: true });
        const handler = TestUtils.findMockCall(window.addEventListener, "keydown")[1];
        const close = jest.fn();
        document.body.addEventListener("ui-library-modal-close", close);

        expect(window.addEventListener).toBeCalled();

        ReactDOM.unmountComponentAtNode(ReactDOM.findDOMNode(component).parentNode);

        expect(window.removeEventListener).toBeCalledWith("keydown", handler);
        expect(close).toBeCalled();
    });

    it("no close event on unmount when component was never opened", function () {
        const component = getComponent({ expanded: false });
        const close = jest.fn();
        const open = jest.fn();
        document.body.addEventListener("ui-library-modal-open", open);
        document.body.addEventListener("ui-library-modal-close", close);

        expect(open).not.toBeCalled();
        component.componentWillUnmount();
        expect(close).not.toBeCalled();
    });

    it("keydown event listener does not trigger when modal is closed", function () {
        const component = getComponent(); //eslint-disable-line
        const handler = TestUtils.findMockCall(window.addEventListener, "keydown")[1];
        const e = {
            target: { parentNode: document.body },
            stopPropagation: jest.fn(),
            preventDefault: jest.fn()
        };

        //expect that the collapsed modal does not process keypress events
        handler(e);
        expect(e.stopPropagation).not.toBeCalled();
    });

    it("keydown event listener triggers when modal is expanded", function () {
        const component = getComponent({ expanded: true }); //eslint-disable-line
        const handler = TestUtils.findMockCall(window.addEventListener, "keydown")[1];
        const e = {
            target: { parentNode: document.body },
            stopPropagation: jest.fn(),
            preventDefault: jest.fn()
        };


        //expect that the collapsed modal does not process keypress events
        handler(e);
        expect(e.stopPropagation).toBeCalled();
    });

    it("emits open and close events", function () {
        const
            openListenerCallback = jest.fn(),
            closeListenerCallback = jest.fn(),
            component = ReactTestUtils.renderIntoDocument(
                <Wrapper
                    type={Modal}
                    expanded={false}
                    onOpen={jest.fn()}
                    onClose={jest.fn()}
                />
            );

        document.body.addEventListener("ui-library-modal-open", openListenerCallback);
        document.body.addEventListener("ui-library-modal-close", closeListenerCallback);

        component._setProps({ expanded: true });
        expect(openListenerCallback).toBeCalled();
        expect(closeListenerCallback).not.toBeCalled();

        component._setProps({ expanded: false });
        expect(closeListenerCallback).toBeCalled();
    });

    it("emits open and close events in IE", function () {
        const UtilsMock = require("../../../util/Utils");
        UtilsMock.isIE = () => { return true; };

        const
            openListenerCallback = jest.fn(),
            closeListenerCallback = jest.fn(),
            component = ReactTestUtils.renderIntoDocument(
                <Wrapper
                    type={Modal}
                    expanded={false}
                    onOpen={jest.fn()}
                    onClose={jest.fn()}
                />
            );

        document.body.addEventListener("ui-library-modal-open", openListenerCallback);
        document.body.addEventListener("ui-library-modal-close", closeListenerCallback);

        component._setProps({ expanded: true });
        expect(openListenerCallback).toBeCalled();
        expect(closeListenerCallback).not.toBeCalled();

        component._setProps({ expanded: false });
        expect(closeListenerCallback).toBeCalled();
    });

    it("Doesn't render body until expanded", function () {
        const htmlDid = "my-content";
        const children = (<div data-id={htmlDid}></div>);
        let component = getComponent({ children: children });
        let content = TestUtils.findRenderedDOMNodeWithDataId(component, htmlDid);

        expect(content).toBeFalsy();

        component = getComponent({ children: children, expanded: true });
        content = TestUtils.findRenderedDOMNodeWithDataId(component, htmlDid);

        expect(content).toBeTruthy();
    });

    it("Test default render", function () {
        const component = getComponent();

        // Expect no modal to be rendered.
        const modals = TestUtils.scryRenderedDOMNodesWithClass(component, "modal");
        expect(modals.length).toEqual(0);
    });

    it("Modal render default data-ids if not provided", function () {
        const component = getComponent({ expanded: true }),
            modal = TestUtils.findRenderedDOMNodeWithClass(component, "modal");

        expect(modal.getAttribute("data-id")).toEqual("modal-button");
    });

    it("Modal render data-ids if provided", function () {
        const did = "myid",
            component = getComponent({ "data-id": did, expanded: true }),
            modal = TestUtils.findRenderedDOMNodeWithDataId(component, did);

        expect(ReactTestUtils.isDOMComponent(modal)).toBeTruthy();
    });

    /*
     * Ensure that a modal set to be initially open
     * via: expanded=true
     * is rendered as such.
     *
     */
    it("Modal initially open", function () {
        const component = getComponent({ initiallyExpanded: true });

        // Expect a single button to be rendered.
        TestUtils.findRenderedDOMNodeWithClass(component, "buttonClass");

        // Expect a single shown modal to be rendered.
        const modal = TestUtils.findRenderedDOMNodeWithClass(component, "modal");
        TestUtils.findRenderedDOMNodeWithClass(modal, "show");
    });

    it("renders in headerless mode", function () {
        const component = getComponent({ expanded: true, showHeader: false });

        // Expect a single button to be rendered.
        TestUtils.findRenderedDOMNodeWithClass(component, "buttonClass");

        // Expect a single shown modal to be rendered.
        const modal = TestUtils.findRenderedDOMNodeWithClass(component, "modal");
        TestUtils.findRenderedDOMNodeWithClass(modal, "show");
        const header = TestUtils.scryRenderedDOMNodesWithClass(modal, "modal-header");

        //no header
        expect(header.length).toBe(0);

        //but close link is here
        TestUtils.findRenderedDOMNodeWithClass(modal, "close-modal");
    });

    it("Cancel tooltip renders and triggers callbacks.", function () {
        const cancelConfirm = jest.fn();
        const cancelDeny = jest.fn();
        const modalParams = {
            expanded: true,
            cancelTooltip: {
                title: "Cancel Confirmation",
                open: false,
                onConfirm: cancelConfirm,
                onCancel: cancelDeny,
                messageText: "Are you sure?",
                confirmButtonText: "Yes",
                cancelButtonText: "No"
            }
        };
        let component = getComponent(modalParams);
        let cancelTooltip = TestUtils.findRenderedDOMNodeWithDataId(component, "modal-button-cancel-tooltip");
        let cancelTooltipContent = TestUtils.findRenderedDOMNodeWithDataId(cancelTooltip, "details-content");

        expect(cancelTooltipContent).toBeFalsy();

        modalParams.cancelTooltip.open = true;
        component = getComponent(modalParams);
        cancelTooltip = TestUtils.findRenderedDOMNodeWithDataId(component, "modal-button-cancel-tooltip");
        cancelTooltipContent = TestUtils.findRenderedDOMNodeWithDataId(cancelTooltip, "details-content");
        expect(cancelTooltip).toBeTruthy();

        const tooltipConfirmBtn = TestUtils.findRenderedDOMNodeWithDataId(component, "modal-button-cancel-confirm-btn"),
            tooltipDenyBtn = TestUtils.findRenderedDOMNodeWithDataId(component, "modal-button-cancel-deny-btn"),
            tooltipTitle = TestUtils.findRenderedDOMNodeWithDataId(component, "details-title"),
            tooltipText = TestUtils.findRenderedDOMNodeWithDataId(component, "modal-button-cancel-tooltip-text");

        ReactTestUtils.Simulate.click(tooltipConfirmBtn);
        expect(cancelConfirm).toBeCalled();
        expect(cancelDeny).not.toBeCalled();

        ReactTestUtils.Simulate.click(tooltipDenyBtn);
        expect(cancelDeny).toBeCalled();

        expect(tooltipConfirmBtn.textContent).toBe(modalParams.cancelTooltip.confirmButtonText);
        expect(tooltipDenyBtn.textContent).toBe(modalParams.cancelTooltip.cancelButtonText);
        expect(tooltipTitle.textContent).toBe(modalParams.cancelTooltip.title);
        expect(tooltipText.textContent).toBe(modalParams.cancelTooltip.messageText);
    });

    it("BodyTitle test", function () {
        const titleText = "This is my title";
        const component =
            shallow(
                <BodyTitle>
                    {titleText}
                </BodyTitle>
            );

        const TitleNode = component.find(".body-title");

        expect(TitleNode.text()).toEqual(titleText);
    });

    it("should render an Unsaved Changes warning popup with ALERT type", function() {
        const component = shallow(<Modal.UnsavedWarningPopup />);

        expect(component.find(Modal).first().props().type).toBe(Modal.Type.ALERT);
    });
});
