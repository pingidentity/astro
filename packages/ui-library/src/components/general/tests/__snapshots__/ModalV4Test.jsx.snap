// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalTest v4 data-id's don't change 1`] = `
Array [
  "modal-button",
  "modal-bg",
  "modal-content",
  "modal-inner-content",
  "modal-header",
  "modal-body",
]
`;
